src/lib/HTTPMethod.js

```diff
@@ -0,0 +1,179 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** A method that is configured to call a service over HTTP. No code or
+  function body is required here, as the actual body is generated to call
+  the remote service. This will always return a promise that supplies the
+  return value of the service call.
+  <p>Overriding by an HTTPMethod is not supported. You can override an
+  HTTPMethod with a normal one.
+  <p>FUTURE: Generalize. */
+foam.CLASS({
+  package: 'foam.net',
+  name: 'HTTPMethod',
+  extends: 'foam.core.Method',
+  requires: [
+    'foam.net.HTTPArgument',
+    'foam.core.Imports',
+  ],
+  imports: [
+    'assert',
+  ],
+
+  properties: [
+    {
+      /** The path prefix. Parameters may add to the path */
+      name: 'path'
+    },
+    {
+      name: 'httpMethod',
+      value: 'GET',
+    },
+    {
+      /** The args to call with, in order */
+      class: 'FObjectArray',
+      name: 'args',
+      of: 'foam.net.HTTPArgument',
+      factory: function() { return []; }
+    },
+    {
+      /** HTTPMethods will always return a Promise, but the Promise will pass
+        along a parameter of the type specified here. */
+      name: 'promisedType',
+      of: 'foam.core.Argument'
+    },
+    [ 'returns', 'Promise' ],
+    {
+      /** The name of the HTTP factory to import at run time. Instances of
+        HTTPMethod on a class will cause the class to import this name, and
+        when called will call hostInstance.yourHttpFactoryName.create() to
+        create a partially filled request object. */
+      name: 'HTTPRequestFactoryName',
+      value: 'HTTPRequestFactory'
+    }
+  ],
+
+  methods: [
+    function installInClass(c) {
+      // add an import for the HTTPRequestFactory on our host class
+
+      // May have many HTTPMethods in a host class, but only do service import once.
+      var existing = c.getAxiomByName(this.HTTPRequestFactoryName);
+      this.assert( existing,
+        "HTTPMethod installInClass did not find an import or property", this.HTTPRequestFactoryName, ".",
+        "Provide one, or set HTTPMethod.HTTPRequestFactoryName to the name of your request factory function.");
+    },
+
+    function installInProto(p) {
+      // generate body to call through to xhr
+      var axiom = this;
+
+      // set up function with correct args, pass them into the
+      // actual implementation, callRemote_()
+      // ALTERNATE: just pass args in and read from arguments
+      var code = "(function "+axiom.name+"_"+"(";
```

Comment:
The only advantage the generated function has is the named arguments, if you happened to debug it, but otherwise we can add type checking or debug help to whatever is generated either way. I'll swap in an args array version